In [212]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder

In [213]:
from numpy import mean
from numpy import std
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.manifold import LocallyLinearEmbedding
from sklearn.linear_model import LogisticRegression
# define dataset
X, y = make_classification(n_samples=1000, n_features=20, n_informative=10, n_redundant=10, random_state=7)
# define the pipeline
steps = [('lle', LocallyLinearEmbedding(n_components=10)), ('m', LogisticRegression())]
model = Pipeline(steps=steps)
# evaluate model
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
n_scores = cross_val_score(model, X, y, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Accuracy: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))

Accuracy: 0.886 (0.028)


In [214]:
adults_df = pd.read_csv("../week_06/adult.csv")
adults_df.columns = ['age', 'workclass', 'sampling weight', 'education', 'education num', 'marital status', 'occupation',
                     'relationship', 'race', 'sex', 'capital gain', 'capital loss', 'hours per week', 'native country', 'income']
adults_df.head()

,age,workclass,sampling weight,education,education num,marital status,occupation,relationship,race,sex,capital gain,capital loss,hours per week,native country,income
0,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
1,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
2,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
3,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
4,37,Private,284582,Masters,14,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K


In [215]:
column_change = ['workclass', 'marital status', 'occupation', 'relationship', 'race', 'sex', 'native country', 'income']
for col in column_change:
    adults_df[col] = adults_df[col].astype(str).str.strip()

In [216]:
adults_df['workclass'] = adults_df['workclass'].astype(str).str.strip()
adults_df['marital status'] = adults_df['marital status'].astype(str).str.strip()


# workclass mapping
workclass_map = {'Private': 0, 'Self-emp-not-inc': 1, 'Local-gove': 2, 'State-gov': 3, 'Self-emp-inc': 4,'Federal-gov': 5, 'Without-pay': 6, 'Never-worked': 7}
adults_df['workclass'] = adults_df['workclass'].map(workclass_map)

# education drop, keep only education number
adults_df = adults_df.drop('education', axis=1)

# marital status mapping
marital_status_map = {'Never-married': 0, 'Married-civ-spouse': 1, 'Divorced': 2, 'Married-spouse-absent': 3, 'Separated': 4, 'Married-AF-spouse': 5, 'Widowed': 6}
adults_df['marital status'] = adults_df['marital status'].map(marital_status_map)

# occupation mapping
occupation_map = {'Adm-clerical': 0, 'Exec-managerial': 1, 'Handlers-cleaners': 2, 'Prof-specialty': 3, 'Other-service': 4, 'Sales': 5, 'Craft-repair': 6,
                  'Transport-moving': 7, 'Farming-fishing': 8, 'Machine-op-inspct': 9, 'Tech-support': 10, 'Protective-serv': 11, 'Armed-Forces': 12}
adults_df['occupation'] = adults_df['occupation'].map(occupation_map)

# relationship mapping
relationship_map = {'Not-in-family': 0, 'Husband': 1, 'Wife': 2,  'Own-child': 3, 'Unmarried': 4, 'Other-relative': 5}
adults_df['relationship'] = adults_df['relationship'].map(relationship_map)

# race mapping
race_map = {'White': 0, 'Black': 1, 'Asian-Pac-Islander': 2, 'Amer-Indian-Eskimo': 3, 'Other': 4}
adults_df['race'] = adults_df['race'].map(race_map)

# sex mapping
sex_map = {'Male': 0, 'Female': 1}
adults_df['sex'] = adults_df['sex'].map(sex_map)

#native country mapping
native_country_map = {'United-States': 1, 'Cuba': 2, 'Jamaica': 3, 'India': 4, '?': 5, 'Mexico': 6, 'South': 7,
                     'Puerto-Rico': 8, 'Honduras': 9, 'England': 10, 'Canada': 11, 'Germany': 12, 'Iran': 13,
                     'Philippines': 14, 'Italy': 15, 'Poland': 16, 'Columbia': 17, 'Cambodia': 18, 'Thailand': 19,
                     'Ecuador': 20, 'Laos': 21, 'Taiwan': 22, 'Haiti': 23, 'Portugal': 24, 'Dominican-Republic': 25,
                     'El-Salvador': 26, 'France': 27, 'Guatemala': 28, 'China': 29, 'Japan': 30, 'Yugoslavia': 31,
                     'Peru': 32, 'Outlying-US(Guam-USVI-etc)':33,'Scotland' :34,'Trinadad&Tobago' :35,
                     'Greece': 36, 'Nicaragua': 37, 'Vietnam': 38, 'Hong': 39, 'Ireland': 40, 'Hungary': 41,
                     'Holand-Netherlands': 42}
adults_df['native country'] = adults_df['native country'].map(native_country_map)

# income country mapping
income_mapping = {'<=50K': 0, '>50K': 1}
adults_df['income']  = adults_df['income'].map(income_mapping)

adults_df.head()

,age,workclass,sampling weight,education num,marital status,occupation,relationship,race,sex,capital gain,capital loss,hours per week,native country,income
0,50,1.0,83311,13,1,1.0,1,0,0,0,0,13,1,0
1,38,0.0,215646,9,2,2.0,0,0,0,0,0,40,1,0
2,53,0.0,234721,7,1,2.0,1,1,0,0,0,40,1,0
3,28,0.0,338409,13,1,3.0,2,1,1,0,0,40,2,0
4,37,0.0,284582,14,1,1.0,2,0,1,0,0,40,1,0


In [217]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from statsmodels.graphics.mosaicplot import mosaic
from sklearn.metrics import ConfusionMatrixDisplay

X = adults_df.drop('income', axis=1)
y = adults_df['income']
X['occupation'] = X['occupation'].fillna(X['occupation'].median())
X['workclass'] = X['workclass'].fillna(X['workclass'].median())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)
X_train

array([[-0.48483403, -0.39618153, -1.44713763, ..., -0.21927284,
         0.36638869, -0.25614242],
       [-1.07206349, -0.39618153, -0.82770666, ..., -0.21927284,
        -0.03866069, -0.25614242],
       [ 0.46941385, -0.39618153,  1.64297682, ..., -0.21927284,
        -0.03866069, -0.25614242],
       ...,
       [-1.43908191, -0.39618153,  0.85624645, ..., -0.21927284,
         0.12335906, -0.25614242],
       [-0.85185244, -0.39618153,  0.20215437, ..., -0.21927284,
        -0.03866069,  0.74400715],
       [ 0.32260648, -0.39618153, -0.68984975, ..., -0.21927284,
        -0.03866069, -0.25614242]], shape=(26048, 13))

In [218]:
adults_new = adults_df.replace('?', np.nan)

y = adults_new['income']
X = adults_new.drop('income', axis=1)

X['occupation'] = X['occupation'].fillna(X['occupation'].median())
X['workclass'] = X['workclass'].fillna(X['workclass'].median())

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

regression = LogisticRegression(random_state=42).fit(X_train, y_train)
y_predicted = regression.predict(X_test)
y_predicted

array([0, 0, 1, ..., 0, 0, 0], shape=(6512,))

In [219]:
regression.score(X_test, y_test)


0.8261670761670762

In [220]:
def precision_and_recall(calc_type: str, tp: int, tn: int,
                          fp: int, fn: int) -> float:
    if calc_type == "precision":
        return tp / (tp + fp)  
    elif calc_type == "recall":
        return tp / (tp + fn)
    else:
        raise ValueError("something happened here. check your inputs")

print(precision_and_recall("precision", 927, 4634, 311, 641))  
print(precision_and_recall("recall", 927, 4634, 311, 641))     

0.7487883683360258
0.5911989795918368


# Linear Discriminant Analysis

In [221]:
from numpy import mean, std
from sklearn.model_selection import cross_val_score, RepeatedStratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

steps = [
    ('lda', LinearDiscriminantAnalysis(n_components=1)),  
    ('logreg', LogisticRegression(random_state=42))
]
model = Pipeline(steps=steps)

cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

n_scores = cross_val_score(model, X_scaled, y, scoring='accuracy', cv=cv, n_jobs=-1)

print('Accuracy after LDA: %.3f (%.3f)' % (mean(n_scores), std(n_scores)))

Accuracy after LDA: 0.813 (0.007)


The accuracy slightly decrease from 0.826 to 0.813 after I apply Linear Discriminant Analysis (a multi-class classification algorithm that can be used for dimensionality reduction). I think it decrease the accuracy slightly because I have already processed the Adult dataset by encoded categorical features and handeled missing values. So when I continue to reduces some features to a lower-dimensional space, it may remove useful signals or patterns, which could lead to the model underfitting. 

Also, this model limiting the number of dimension to 1. I think that by limiting it to 1 in a binary classification problem, LDA can reduces the data to the single most discriminative direction, which removes less informative variation and can reduce noise. However, it may also discard some useful signals if the dataset is already pretty well processed.